File: Algorithms/375/guess-number-higher-or-lower-ii.swift
class Solution {
    func getMoneyAmount(_ n: Int) -> Int {
        if n < 2 {
            return 0
        }
        var payAry = Array.init(repeating: Array.init(repeating: 0, count: n + 5), count: n + 5)
        
        for len in 2...n {
            for st in 1...n {
                let ed = st + len-1
                if ed > n {
                    continue
                }
                
                var pay = Int.max
                for i in st...ed {
                    let tmp = i + max(payAry[st][i - st], payAry[i+1][ed - i])
                    pay = min(pay, tmp)
                }
                payAry[st][len] = pay
            }
        }
        
        return payAry[1][n]
    }
}

File: Algorithms/[1, 1k)/375/guess-number-higher-or-lower-ii.go
package main

import (
	"fmt"
)

func getMoneyAmount(n int) int {
	if n < 2 {
		return 0
	}

	payAry := make([][]int, n+5)
	for i := 0; i < n+5; i++ {
		payAry[i] = make([]int, n+5)
	}

	for len := 2; len <= n; len++ {
		for st := 1; st <= n; st++ {
			ed := st + len - 1
			if ed > n {
				continue
			}

			pay := 0x7FFFFF
			for i := st; i <= ed; i++ {
				tmp := i + maxInt(payAry[st][i-st], payAry[i+1][ed-i])
				pay = minInt(pay, tmp)
			}
			payAry[st][len] = pay
		}
	}

	return payAry[1][n]
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	for i := 1; i <= 20; i++ {
		fmt.Println(getMoneyAmount(i))
	}
	fmt.Println(getMoneyAmount(100))
}
